src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -869,71 +938,152 @@ internal static unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase(ref char fi
-            return GetNonRandomizedHashCodeOrdinalIgnoreCaseSlow(hash1, hash2, ref firstChar, length);
+            return GetNonRandomizedHashCodeOrdinalIgnoreCaseSlow(hash1, hash2, this.AsSpan(Length - length));
+        }
 
-            static int GetNonRandomizedHashCodeOrdinalIgnoreCaseSlow(uint hash1, uint hash2, ref char firstChar, int length)
-            {
-                char[]? borrowedArr = null;
-                // Important: leave an additional space for '\0'
-                Span<char> scratch = (uint)length < 256 ?
-                    stackalloc char[256] :
-                    (borrowedArr = ArrayPool<char>.Shared.Rent(length + 1));
+        internal static unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase(ReadOnlySpan<char> span)
+        {
+            uint hash1 = (5381 << 16) + 5381;
+            uint hash2 = hash1;
 
-                int charsWritten = Ordinal.ToUpperOrdinal(new ReadOnlySpan<char>(ref firstChar, length), scratch);
-                Debug.Assert(charsWritten == length);
-                scratch[length] = '\0';
+            uint p0, p1;
+            int length = span.Length;
 
-                const uint NormalizeToLowercase = 0x0020_0020u;
+            fixed (char* src = &MemoryMarshal.GetReference(span))
+            {
+                uint* ptr = (uint*)src;
 
-                // Duplicate the main loop, can be removed once JIT gets "Loop Unswitching" optimization
-                fixed (char* src = scratch)
+                LengthSwitch:
+                switch (length)
                 {
-                    uint* ptr = (uint*)src;
-                    while (length > 2)
-                    {
-                        length -= 4;
-                        hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (ptr[0] | NormalizeToLowercase);
-                        hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[1] | NormalizeToLowercase);
-                        ptr += 2;
-                    }
+                    default:
+                        do
+                        {
+                            p0 = Unsafe.ReadUnaligned<uint>(ptr);
+                            p1 = Unsafe.ReadUnaligned<uint>(ptr + 1);
+                            if (!Utf16Utility.AllCharsInUInt32AreAscii(p0 | p1))
+                            {
+                                goto NotAscii;
+                            }
+
+                            length -= 4;
+                            hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (p0 | NormalizeToLowercase);
+                            hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (p1 | NormalizeToLowercase);
+                            ptr += 2;
+                        }
+                        while (length >= 4);
+                        goto LengthSwitch;
+
+                    case 3:
+                        p0 = Unsafe.ReadUnaligned<uint>(ptr);
+                        p1 = *(char*)(ptr + 1);
+#if BIGENDIAN
```

Comment:
nit: I think we've been trying to use `#ifdef` a bit less in places and so using `!BitConverter.IsLittleEndian` might be preferred, especially given the linker/trimmer optimizes it down the same.

On a separate note, it might be nice if we also had `BitConverter.IsBigEndian` so that the context is clearer around such APIs.

Reply:
Ok, I can fix that up separately.

Reply:
> especially given the linker/trimmer optimizes it down the same.

not really the same https://github.com/dotnet/runtime/issues/83169 but I guess it's not a big deal